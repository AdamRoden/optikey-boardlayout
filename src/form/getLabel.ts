const getLabel = (label: string) => {
  const i = label.indexOf("]")
  return i === -1 ? label : label.substring(i + 2)
}

export default getLabel
