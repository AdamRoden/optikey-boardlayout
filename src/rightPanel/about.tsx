import { AppBar, Button, Dialog, DialogContent, IconButton, Slide, Toolbar } from "@material-ui/core"
import ExpansionPanel from "@material-ui/core/ExpansionPanel"
import ExpansionPanelDetails from "@material-ui/core/ExpansionPanelDetails"
import ExpansionPanelSummary from "@material-ui/core/ExpansionPanelSummary"
import Typography from "@material-ui/core/Typography"
import CloseIcon from "@material-ui/icons/Close"
import { observable } from "mobx"
import { observer } from "mobx-react-lite"
import * as React from "react"
import licenses from "../util/licenses"

const classes: any = {
  appBar: {
    position: "relative"
  },
  flex: {
    flex: 1
  }
}
function Transition(props: any) {
  return <Slide direction="up" {...props} />
}

const isOpen = observable.box(false)
const handleClose = () => {
  isOpen.set(false)
}
const handleOpen = () => isOpen.set(true)

const AboutPanel = observer(() => {
  return (
    <ExpansionPanel defaultExpanded>
      <ExpansionPanelSummary>
        <Typography>About</Typography>
      </ExpansionPanelSummary>
      <ExpansionPanelDetails>
        <Typography>
          Written with ❤️ by <a>hizoul</a><br/>
          Special thanks to&nbsp;
          <a href="https://github.com/JuliusSweetland">Julius Sweetland</a>
          &nbsp;for writing&nbsp;
          <a href="https://github.com/OptiKey/OptiKey">
            OptiKey
          </a> and&nbsp;
          <a
            onClick={handleOpen}
            href="#"
          >
            all other authors of open source libraries
          </a>
          &nbsp;that make this small tool possible
        </Typography>
      </ExpansionPanelDetails>
      <Dialog
        fullScreen
        open={isOpen.get()}
        onClose={handleClose}
        TransitionComponent={Transition}
      >
        <AppBar className={classes.appBar}>
          <Toolbar>
            <Typography variant="h6" color="inherit" className="flex1">
              Third Party Licenses
            </Typography>
            <IconButton color="inherit" onClick={handleClose} aria-label="Close">
              <CloseIcon />
            </IconButton>
          </Toolbar>
        </AppBar>
        <DialogContent>
          <Typography style={{margin: "5rem 0rem"}}>
            {licenses.split("\n").map((s) => <div key={s}>{s}</div>)}
          </Typography>
        </DialogContent>
      </Dialog>
    </ExpansionPanel>
  )
})

export default AboutPanel
