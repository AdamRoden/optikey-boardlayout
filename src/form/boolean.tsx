import Checkbox from '@material-ui/core/Checkbox'
import FormControlLabel from '@material-ui/core/FormControlLabel'
import { getMapToFromProps, IFieldProps, useFieldWithValidation } from "@xpfw/form"
import { observer } from "mobx-react-lite"
import * as React from "react"
import getLabel from "./getLabel"

const changeWrapper = (fieldHelper: {setValue: Function, value: any}) => {
  return (e: any) => {
    fieldHelper.setValue(!fieldHelper.value)
  }
}

const BooleanField: React.FunctionComponent<IFieldProps & {
  className?: string
  placeholder?: string
}> = observer((props) => {
  const mapTo = getMapToFromProps(props)
  const fieldHelper = useFieldWithValidation(props.schema, mapTo, props.prefix)
  const setBool = changeWrapper(fieldHelper)
  return (
    <FormControlLabel
        control={
          <Checkbox
            checked={fieldHelper.value}
            onChange={setBool}
            value={fieldHelper.value}
          />
        }
        label={getLabel(mapTo)}
    />
  )
})

export default BooleanField
export {
  changeWrapper
}