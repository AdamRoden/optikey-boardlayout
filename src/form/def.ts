import { ComponentRegistry, ExtendedJSONSchema } from "@xpfw/form"
import { registerComponents } from "@xpfw/form-web"
import BooleanField from "./boolean"
import SelectField from "./select"
import TextField from "./text"

registerComponents()
ComponentRegistry.registerComponent("number", TextField)
ComponentRegistry.registerComponent("string", TextField)
ComponentRegistry.registerComponent("boolean", BooleanField)
ComponentRegistry.registerComponent("string", SelectField, "select")

const nameSchema: ExtendedJSONSchema = {
  title: "Name", type: "string"
}

const heightSchema: ExtendedJSONSchema = {
  title: "Height", type: "number"
}

const symbolSchema: ExtendedJSONSchema = {
  title: "SymbolMargin", type: "number"
}

const borderSchema: ExtendedJSONSchema = {
  title: "BorderThickness", type: "string"
}

const rowsSchema: ExtendedJSONSchema = {
  title: "Grid.Rows", type: "number"
}

const colsSchema: ExtendedJSONSchema = {
  title: "Grid.Cols", type: "number"
}

const keyboardPrefix = "Keyboard"
const hoverKey = `hover`
const dragKey = `drag`
const activeKey = `active`

export {
  nameSchema, heightSchema, borderSchema, symbolSchema, rowsSchema, colsSchema, keyboardPrefix,
  activeKey, dragKey, hoverKey
}
