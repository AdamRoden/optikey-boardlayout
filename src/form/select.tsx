import FormControl from '@material-ui/core/FormControl'
import FormGroup from '@material-ui/core/FormGroup'
import FormHelperText from '@material-ui/core/FormHelperText'
import InputLabel from '@material-ui/core/InputLabel'
import MenuItem from '@material-ui/core/MenuItem'
import Select from '@material-ui/core/Select'
import CheckBoxIcon from '@material-ui/icons/CheckBox'
import { getMapToFromProps, IFieldProps, JSONSchemaDefinition, memo, useFieldWithValidation } from "@xpfw/form"
import { get, isFunction } from "lodash"
import { observer } from "mobx-react-lite"
import * as React from "react"
import getLabel from "./getLabel";

const SelectField: React.FunctionComponent<IFieldProps & {
  className?: string
  placeholder?: string
}> = observer((props) => {
  const mapTo = getMapToFromProps(props)
  const fieldHelper = useFieldWithValidation(props.schema, mapTo, props.prefix)
  let selOpts = get(props, "schema.selectOptions", [])
  if (isFunction(selOpts)) {
    selOpts = selOpts(fieldHelper.value, props.schema, props)
  }
  const options = selOpts.map((option: any) => {
    return (
      <MenuItem key={option.value} value={option.value}>
        {option.label}
      </MenuItem>
    )
  })
  return (
    <FormControl fullWidth>
      <InputLabel>{getLabel(mapTo)}</InputLabel>
        <Select
          onChange={(e) => {
            const t: any = get(e, "nativeEvent.target")
            fieldHelper.setValue(t.getAttribute("data-value"))
          }}
          value={fieldHelper.value}
        >
          {options}
        </Select>
     </FormControl>
  )
})

export default SelectField
