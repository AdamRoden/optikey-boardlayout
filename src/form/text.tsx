import TextFiel from '@material-ui/core/TextField'
import { getMapTo, getMapToFromProps, IFieldProps, JSONSchemaDefinition, memo, useFieldWithValidation } from "@xpfw/form"
import { get } from "lodash"
import { observer } from "mobx-react-lite"
import * as momentA from "moment"
import * as React from "react"
const MTextField: any = TextFiel
const TextField: React.FunctionComponent<IFieldProps & {
  className?: string
  placeholder?: string
}> = observer((props) => {
  const format = get(props, "schema.format")
  const isDate = format === "date" || format === "date-time" || format === "time"
  const fieldHelper = useFieldWithValidation(props.schema, getMapToFromProps(props), props.prefix, {
    valueEventKey: "nativeEvent.target.value"
  })
  const fieldType = get(props, "schema.type")
  const value = fieldHelper.value
  let type = "text"
  let min
  let max
  let step
  const onChange = fieldHelper.setValue
  if (fieldType === "number") {
    type = "number"
    min = get(props, "schema.minimum")
    max = get(props, "schema.maximum")
    step = get(props, "schema.step")
  }
  return (
    <MTextField
      type={type}
      label={getMapTo(props.schema, props.mapTo)}
      id={get(props, "id")}
      className={get(props, "className")}
      value={value}
      inputProps={{step, min, max}}
      onChange={onChange}
      placeholder={props.placeholder}
      fullWidth
    />
  )
})

export default TextField
