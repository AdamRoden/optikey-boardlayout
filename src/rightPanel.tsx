import { ComponentRegistry, ExtendedJSONSchema, FormStore, SharedField } from "@xpfw/form"
import { registerComponents, TextField } from "@xpfw/form-web"
import * as React from "react"

registerComponents()
ComponentRegistry.registerComponent("number", TextField)

const nameSchema: ExtendedJSONSchema = {
  title: "Name", type: "string"
}

const heightSchema: ExtendedJSONSchema = {
  title: "Height", type: "number"
}
const symbolSchema: ExtendedJSONSchema = {
  title: "SymbolMargin", type: "number"
}
const borderSchema: ExtendedJSONSchema = {
  title: "BorderThickness", type: "string"
}
const rowsSchema: ExtendedJSONSchema = {
  title: "Grid.Rows", type: "number"
}
const colsSchema: ExtendedJSONSchema = {
  title: "Grid.Cols", type: "number"
}
const keyboardPrefix = "Keyboard"

const RightPanel: React.FunctionComponent<any> = () => (
  <div>
    hi panep
    <SharedField schema={nameSchema} prefix={keyboardPrefix} />
    <SharedField schema={heightSchema} prefix={keyboardPrefix} />
    <SharedField schema={borderSchema} prefix={keyboardPrefix} />
    <SharedField schema={rowsSchema} prefix={keyboardPrefix} />
    <SharedField schema={colsSchema} prefix={keyboardPrefix} />
    <SharedField schema={symbolSchema} prefix={keyboardPrefix} />
  </div>
)

export default RightPanel
