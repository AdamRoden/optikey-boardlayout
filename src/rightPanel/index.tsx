import Button from "@material-ui/core/Button"
import ExpansionPanel from "@material-ui/core/ExpansionPanel"
import ExpansionPanelDetails from "@material-ui/core/ExpansionPanelDetails"
import ExpansionPanelSummary from "@material-ui/core/ExpansionPanelSummary"
import Typography from "@material-ui/core/Typography"
import ExpandMoreIcon from "@material-ui/icons/ExpandMore"
import { SharedField } from "@xpfw/form"
import * as React from "react"
import {
  borderSchema, colsSchema, heightSchema, keyboardPrefix, nameSchema, rowsSchema, symbolSchema
} from "../form/def"
import downloadKeyboard, { copyToClipboard } from "../util/download"
import AboutPanel from "./about"
import KeyboardImporter from "./import"
import KeyEditor from "./keyEditor"

const RightPanel: React.FunctionComponent<any> = () => (
  <div className="rightPanel">
    <div className=" row">
      <Button variant="outlined" color="primary" fullWidth onClick={downloadKeyboard}>download xml</Button>
      <Button variant="outlined" color="primary" fullWidth onClick={copyToClipboard}>copy xml to clipboard</Button>
    </div>
    <KeyboardImporter />
    <ExpansionPanel defaultExpanded={false}>
      <ExpansionPanelSummary  expandIcon={<ExpandMoreIcon />}>
        <Typography>Global Settings</Typography>
      </ExpansionPanelSummary>
      <ExpansionPanelDetails>
        <div className="center" style={{flexDirection: "column", flex: 1}}>
          <SharedField schema={nameSchema} prefix={keyboardPrefix} />
          <SharedField schema={heightSchema} prefix={keyboardPrefix} />
          <SharedField schema={borderSchema} prefix={keyboardPrefix} />
          <SharedField schema={rowsSchema} prefix={keyboardPrefix} />
          <SharedField schema={colsSchema} prefix={keyboardPrefix} />
          <SharedField schema={symbolSchema} prefix={keyboardPrefix} />
        </div>
      </ExpansionPanelDetails>
    </ExpansionPanel>
    <KeyEditor />
    <AboutPanel />
  </div>
)

export default RightPanel
