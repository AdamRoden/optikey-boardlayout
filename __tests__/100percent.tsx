import * as React from "react"
import { changeWrapper } from "../src/form/boolean"
import SelectField, { changeWrapperSelect } from "../src/form/select"
import renderSnapshot from "../src/testUtil/renderSnapshot"
import { Transition } from "../src/rightPanel/about"

test("100percent", () => {
  // test boolean change function
  const fieldHelper: any = {value: undefined}
  fieldHelper.setValue = (newVal: any) => fieldHelper.value = newVal
  let changer = changeWrapper(fieldHelper)
  changer(null)
  expect(fieldHelper.value).toBe(true)
  changer(null)
  expect(fieldHelper.value).toBe(false)

  const changerSelect = changeWrapperSelect(fieldHelper)
  changerSelect({nativeEvent: {target: {getAttribute: () => "customVal"}}})
  expect(fieldHelper.value === "customVal").toBeTruthy()
  const selectSchema: any = {
    title: "my select with func",
    selectOptions: () => ["my ", "select", "function", "args"]
  }
  renderSnapshot(<SelectField schema={selectSchema} />, "select field with function args")

  renderSnapshot(<Transition/>, "transition render")
})