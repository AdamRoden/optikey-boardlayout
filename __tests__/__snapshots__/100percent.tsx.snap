// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`100percent: select field with function args 1`] = `
<div
  className="MuiFormControl-root MuiFormControl-fullWidth"
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
    data-shrink={false}
  >
    my select with func
  </label>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
  >
    <div
      aria-haspopup="listbox"
      aria-labelledby=" "
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      role="button"
      tabIndex={0}
    >
      <span
        dangerouslySetInnerHTML={
          Object {
            "__html": "&#8203;",
          }
        }
      />
    </div>
    <input
      onAnimationStart={[Function]}
      type="hidden"
    />
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root MuiSelect-icon"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
  </div>
</div>
`;

exports[`100percent: transition render 1`] = `
<div>
  slide 
  {
  "direction": "up"
}
</div>
`;
